pkg/virtctl/vmexport/vmexport.go

```diff
@@ -0,0 +1,650 @@
+				if format.Format == exportv1.KubeVirtGz || format.Format == exportv1.ArchiveGz {
+					downloadUrl = format.Url
+					break
+				} else if format.Format == exportv1.KubeVirtRaw {
+					downloadUrl = format.Url
+				}
+			}
+		}
+	}
+
+	if downloadUrl == "" {
+		return "", fmt.Errorf("Unable to get a valid URL from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
+	}
+
+	return downloadUrl, nil
+}
+
+// waitForVirtualMachineExport waits for the VirtualMachineExport status and external links to be ready
+func waitForVirtualMachineExport(client kubecli.KubevirtClient, vmeInfo *VMExportInfo, interval, timeout time.Duration) error {
+	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		vmexport, err := getVirtualMachineExport(client, vmeInfo)
+		if err != nil {
+			return true, err
+		}
+
+		if vmexport == nil {
+			return true, err
+		}
+
+		if vmexport.Status == nil {
+			return false, nil
+		}
+
+		if vmexport.Status.Phase != exportv1.Ready {
+			fmt.Printf("Waiting for VM Export %s status to be ready...\n", vmeInfo.Name)
+			return false, nil
+		}
+
+		if vmexport.Status.Links == nil || vmexport.Status.Links.External == nil {
+			fmt.Printf("Waiting for VM Export %s external links to be available...\n", vmeInfo.Name)
+			return false, nil
+		}
+
+		fmt.Printf("Processing completed successfully\n")
+		return true, nil
+	})
+
+	return err
+}
+
+// handleHTTPRequest generates the GET request with proper certificate handling
+func handleHTTPRequest(client kubecli.KubevirtClient, vmexport *exportv1.VirtualMachineExport, downloadUrl string, insecure bool) (*http.Response, error) {
+	token, err := getTokenFromSecret(client, vmexport)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create new certPool and append our external SSL certificate
+	cert := vmexport.Status.Links.External.Cert
+	roots := x509.NewCertPool()
+	roots.AppendCertsFromPEM([]byte(cert))
+	transport := &http.Transport{
+		TLSClientConfig: &tls.Config{RootCAs: roots},
+	}
+	httpClient := httpClientCreatorFunc(transport, insecure)
+
+	// Generate and do the request
+	req, _ := http.NewRequest("GET", downloadUrl, nil)
+	req.Header.Set(exportTokenHeader, token)
+
+	return httpClient.Do(req)
+}
+
+// getHTTPClient assigns the default, non-mocked HTTP client
+func getHTTPClient(transport *http.Transport, insecure bool) *http.Client {
+	if insecure == true {
+		transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: insecure,
+			},
+		}
+	}
+
+	client := &http.Client{Transport: transport}
+	return client
+}
+
+// copyFileWithProgressBar serves as a wrapper to copy the file with a progress bar
+func copyFileWithProgressBar(output *os.File, resp *http.Response) error {
+	barTemplate := fmt.Sprintf(`{{ "Downloading file:" }} {{counters . }} {{ cycle . %s }} {{speed . }}`, progressBarCycle)
+
+	// start bar based on our template
+	bar := pb.ProgressBarTemplate(barTemplate).Start(0)
```

Comment:
Curious how the progress bar works with archive format?  The total size of the archive is unknown at the start of the request.

Reply:
So I initially wanted to use the file length for the progress bar but encountered that problem: The file size was unknown, so the bar wasn't behaving very well. I managed it by using a custom progress bar that works as a spinner and displays a cycle until the file is downloaded, independently of the size.